tests/test_ddc_overwrite.c

```diff
@@ -0,0 +1,40 @@
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "cheriintrin.h"
+
+/* This test atempts to circumvent the DDC compartmentalization feature by
+ * using the existing PCC as a temporary DDC to access a secret pointer beyond
+ * the original bounds of the compartment. The address of the secret is
+ * expected to be passed as `secret_addr`, and the function will return the
+ * value at that address. If the given `secret_addr` is out of the original DDC
+ * bounds, and this test does not fail, then we have broken the expected
+ * security guarantees of our DDC compartmentalization approach.
+ */
+
+int
+test_overwrite(unsigned long long secret_addr)
+{
+    uintptr_t addr = (uintptr_t) secret_addr;
+    void* __capability curr_pcc = cheri_pcc_get();
+    void* __capability old_ddc = cheri_ddc_get();
+    int secret;
+    asm volatile("msr ddc, %[cap]\n\t"
+                 "ldr x0, [%[addr]]\n\t"
+                 "msr ddc, %[old_cap]\n\t"
+                 "mov %w[val], w0"
```

Comment:
* Setting the DDC is probably inefficient compared with just deriving a new capability from PCC directly (`cvtp`).
* If you only actually want 32 bits, then do you need to load 64? (Maybe you do, to stress bounds or something, but it doesn't appear that that's the case.)

Consider:

```
cvtp c0, [%[addr]]
ldr %w[secret], [c0]
```

Reply:
The 32 bits, do those come from using `ldr x0`? Was the suggestion here to have used `ldr w0` (which has been removed in the updated code)?

Updated the code in https://github.com/capablevms/CHERI-ELF-comp/pull/12/commits/42ee892cab7a9700319e993dea5b9f85b78ff504.

Reply:
> The 32 bits, do those come from using ldr x0?

`ldr x0, ...` loads a 64-bit value, which (in the version I reviewed) was then truncated with `mov %w[val], w0`. It will abort if you can't load the whole 64-bit value, even though you only return the result.
`ldr %w[val], ...` just loads a 32-bit value in the first place, so you don't need to truncate it.